### ISSUE TITLE
Possible segfault with ccall

### ISSUE THREAD

**Issue by Lecrapouille** [JuliaLang/julia#41890]
Following https://discourse.julialang.org/t/why-this-code-does-a-segfault/66394
In a particular case, I guess Julia is doing a segfault instead of throwing an error (while doing it correctly in other case).

I'm on a 64-bits system.

## foo.c
```c
#include <stdio.h>

struct mystruct
{
    ssize_t a;
    ssize_t b;
};

typedef struct mystruct mystruct_t;

int recvstruct(mystruct_t* st)
{
    printf("C %u, %u\n", st->a, st->b);
    return 0;
}

mystruct_t* updatestruct(mystruct_t* st)
{
    st->a = 10;
    st->b = 11;
    return st;
}
```

Compiled with:
```
gcc foo.c -fpic -shared -o libfoo.so
```

## Julia.jl

This code is working
```
struct mystruct
    a::Int
    b::Int
end

function let_c_print_struct2(v::mystruct)
    ccall((:recvstruct, "./libfoo.so"), Int, (Ref{mystruct},), v)
end

function let_c_update_struct2()
    st = mystruct(5, 6)
    ret = ccall((:updatestruct, "./libfoo.so"), Ref{mystruct}, (Ref{mystruct},), st)
    let_c_print_struct2(ret)

    @show typeof(ret)
    ret
end
```

Execute:
```
julia> ret = let_c_update_struct2()
C 10, 11
typeof(ret) = mystruct
mystruct(10, 11)

julia> let_c_print_struct2(ret)
C 10, 11
0
```

But in `let_c_update_struct2()` if I used `ret[]` instead of `ret` this produce a segfault while the `typeof` is detected as `mystruct`.
But now if I do that:

```
julia> foo = mystruct(5, 6)
mystruct(5, 6)

julia> foo[]
ERROR: MethodError: no method matching getindex(::mystruct)
Stacktrace:
 [1] top-level scope
   @ REPL[7]:1
```
Julia detects correctly, this is not a Ref. But so why inside `let_c_update_struct2` julia did not throw the same error ?

**Comment by StefanKarpinski** [comment]
This is almost certainly a misuse of ccall, not a bug, so I'm closing this. If the discussion on discourse turns up evidence of a bug, feel free to ask for this to be reopened.

**Comment by vtjnash** [comment]
In particular, the user error seems to have been solved in the discourse thread

**Comment by Lecrapouille** [comment]
> In particular, the user error seems to have been solved in the discourse thread

I solved my problem, but since with the new solution I'm no longer use the returned pointer, I did not check it back. But my issue was that I was defining `Ref` instead `Ptr` inside the definition of `ccall` and Julia was garbaging some static C memory (referenced by my pointers), which ended badly. I guess in this GitHub ticket what is happening is the same: the returned pointer was Ref instead of Ptr and therefore `typeof(ret)` returned `mystruct` (but not `Ref{mystruct}` because of `Base.cconvert`) and when I was trying to do `typeof(ret[])` Julia tried to refer to a failed garbaged memory and ended. I may wrong, but I think `Base.cconvert` does a data copy from Ref to non-Ref then call a free().

The documentation is not very beginner-friendly on this point. May be adding some charts to help to understand on what is happening on the memory could help.

PS: concerning a remark about `Cint` aliasing `Int32` I think this dangerous since in 64-bits archi int == ssize_t == int64_t
